/*
 * Copyright 2022-2026 Leonard Lemke
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.lemke.sudoku.domain

import de.lemke.sudoku.data.database.SudokusRepository
import de.lemke.sudoku.domain.model.Sudoku
import io.kotest.core.spec.style.ShouldSpec
import io.mockk.coVerify
import io.mockk.mockk
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.UnconfinedTestDispatcher

@OptIn(ExperimentalCoroutinesApi::class)
class SaveSudokuUseCaseTest : ShouldSpec(
    {
        val sudokusRepository = mockk<SudokusRepository>(relaxUnitFun = true)
        val useCase = SaveSudokuUseCase(sudokusRepository, UnconfinedTestDispatcher())
        val sudoku = mockk<Sudoku>()

        should("saves with onlyUpdate = false by default") {
            useCase(sudoku)

            coVerify(exactly = 1) { sudokusRepository.saveSudoku(sudoku, false) }
        }

        should("saves with onlyUpdate = true when explicitly requested") {
            useCase(sudoku, onlyUpdate = true)

            coVerify(exactly = 1) { sudokusRepository.saveSudoku(sudoku, true) }
        }
    },
)
